### DISCUSSION TITLE
[SDK] EIP-712 gas price and viem: gasPrice or EIP-1559?

### DISCUSSION THREAD

**Discussion by cesarenaldi** [zkSync-Community-Hub/zksync-developers#808] [SDKs] [answered]
### Team or Project

Lens Network

### Select the SDK

JavaScript

### Environment

Testnet

### Issue Description

From past conversations, we have been using pre-EIP1559 `gasPrice` field to populate gas price for EIP712 txs. This seems to not work when tx are send through viem.

```
Execution reverted with reason: max fee per gas less than block base fee.
```

Given the error message suggests a response from the node, we are not sure if this is an issue w/ the request or the client.

We do send EIP712 transactions in the same way through ethers.js and allow-rs without any issue.
A quick hack to use `maxFeePerGas` and `maxPriorityFeePerGas` seems to fix the issue also with viem.

### Expected Behavior

This ticket is more to clarify what is the recommended way to handle gas prices.
We are getting conflicting information, and we couldn't find an official doc that make this clear.

Should we use legacy `gasPrice` or EIP-1559 `maxFeePerGas` and `maxPriorityFeePerGas`?

### Code Example

```ts
import { chains } from '@lens-network/sdk/viem';
import { parseAbi, Hex, createWalletClient, http, encodeFunctionData } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { getGasPrice } from 'viem/actions';
import { eip712WalletActions, getGeneralPaymasterInput, sendEip712Transaction } from 'viem/zksync';

const chain = chains.testnet;

const account = privateKeyToAccount(process.env.PRIVATE_KEY as Hex);

const walletClient = createWalletClient({
  account,
  chain,
  transport: http(),
}).extend(eip712WalletActions());

const abi = parseAbi([
  "function ping()"
]);

const hash = await sendEip712Transaction(walletClient, {
  account,
  to: '0x06a71429d153026a0F3bAdf481c216FDfe2d0629',
  data: encodeFunctionData({
    abi: abi,
    functionName: 'ping'
  }),
  paymaster: '0xE6b37F099439ED1Eb38a97b909E9AD0ad74270Af',
  paymasterInput: getGeneralPaymasterInput({
    innerInput: '0x',
  }),
  gasPrice: await getGasPrice(walletClient)
})
```

**Comment by alhenaai** [comment]
In the ZKsync Era, while the EIP-1559 transaction format is supported, the `maxFeePerGas` and `maxPriorityFeePerGas` parameters are not utilized. This means that even though the transaction format allows for these fields, they do not play a role in the transaction fee calculation within ZKsync Era. Instead, the system uses a different mechanism to handle transaction fees, which does not rely on these EIP-1559 specific parameters.

Given the error you encountered ("max fee per gas less than block base fee"), it suggests that the node (viem in this case) is expecting EIP-1559 style gas parameters but is not handling them as expected in the context of ZKsync Era. This could be due to a discrepancy in how the viem node interprets or expects these parameters compared to the standard ZKsync Era node behavior.

For transactions on ZKsync Era, it's recommended to use the legacy `gasPrice` field if that aligns with the expected behavior of the network and node you are interacting with. However, if using `maxFeePerGas` and `maxPriorityFeePerGas` resolves the issue with viem and is compatible with your transaction requirements, you may continue using them as a workaround.

For more detailed information on how gas is handled in ZKsync Era, you can refer to the [ZKsync documentation on transaction lifecycle](https://docs.zksync.io/zk-stack/concepts/transaction-lifecycle).

For further discussions and developer support, you might find it helpful to visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
> Should we use legacy gasPrice or EIP-1559 maxFeePerGas and maxPriorityFeePerGas?

From what I've found so far you'll want to use EIP-1559 `maxFeePerGas` and `maxPriorityFeePerGas`

**Reply by bxpana** [reply]
@cesarenaldi does that work for you or do you need to use `gasPrice` ?

**Reply by cesarenaldi** [reply]
I don't 'need' to use `gasPrice` just need to understand what's the official way given the discrepancy between tools/docs. Happy to adopt `maxFeePerGas` and `maxPriorityFeePerGas` if that's the way.

**Comment by danijelTxFusion** [comment]
In `zksync-ethers`, you can use `gasPrice` in an EIP-712 transaction because it is converted to `maxFeePerGas` [internally](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/signer.ts#L111). However, in `viem`, the `sendEip712Transaction` function does not support the `gasPrice` parameter, and TypeScript will give you a compilation error with the following message:

```ts
sendEip712Transaction(wallet, {
    account,
    to: '0xa61464658AfeAf65CccaaFD3a512b69A83B77618',
    value: 1_000_000_000n,
    paymaster: '0xa222f0c183AFA73a8Bc1AFb48D34C88c9Bf7A174',
    paymasterInput: getApprovalBasedPaymasterInput({
        minAllowance: 1n,
        token: '0x841c43Fa5d8fFfdB9efE3358906f7578d8700Dd4',
        innerInput: new Uint8Array(),
    }),
    gasPrice: await getGasPrice(wallet)    // <-- not supported
})
```
Error:
```text
Types of property 'gasPrice' are incompatible.
Type 'bigint' is not assignable to type 'undefined'.
```
In `viem`, you should use the `maxFeePerGas` parameter instead of `gasPrice`, as it does not automatically convert `gasPrice` to `maxFeePerGas` like `zksync-ethers` does.

**Reply by cesarenaldi** [reply]
@danijelTxFusion what version of viem are you running. The type issue is not surfacing on mine I am on 2.21.53.

Regardless of the field, the typical EIP-1559 gas price is usually more involved than 'legacy gas price'. You know requires to consider gas prices of the last few blocks etc.

What is the official recommendation on gas price estimation technique for ZKsync L2 chains? Can you please advise?

**Reply by danijelTxFusion** [reply, marked as answer]
The best approach for gas and price estimation is to use [zks_estimateFee](https://docs.zksync.io/build/api-reference/zks-rpc#zks_estimatefee) endpoint. Both [zksync-ethers](https://sdk.zksync.io/js/ethers/api/v6/providers/provider#estimatefee) and [viem](https://viem.sh/zksync/actions/estimateFee) provide `estimateFee` method.
